### PR TITLE
Feature/worldgen

### PR DESCRIPTION
**Pull request by IchHabeHunger54**
This PR:
- moves the worldgen stuff over to `DeferredRegister`s
- adds rare moonstone meteorites, which are the source of moonstone
- adds sunstone ore near the nether's lava oceans
- adds uses for moonstone and sunstone in spell recipes and altar building
- adds the necessary compendium entries

### CODE REVIEW

File: src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMEnglishLanguageProvider.java [hunk 1]
```diff
@@ -241,7 +241,7 @@ protected void addTranslations() {
...
         skillTranslation(AMSpellParts.BLINK.getId(), "Blink", "Well, I'm out.", "components", "You can teleport a short distance directly forward the way you are facing.$(br2)Blink can take you through solid walls, but will make every effort to ensure you don't get stuck in one.");
-        skillTranslation(AMSpellParts.BLIZZARD.getId(), "Blizzard", "Snow. Lots of snow.", "components", "You have learned to summon a fearsome blizzard, which will slow and damage any entities in its radius.$(br2)Blizzard has a built-in $(l:shapes/aoe)AoE$() to it.$(br2)Though blizzard will harm all entities in its radius, it will never harm its caster.");
+        skillTranslation(AMSpellParts.BLIZZARD.getId(), "Blizzard", "Snow. Lots of snow.", "components", "You have learned to summon a fearsome blizzard, which will slow and damage any entities in its craterRadius.$(br2)Blizzard has a built-in $(l:shapes/aoe)AoE$() to it.$(br2)Though blizzard will harm all entities in its craterRadius, it will never harm its caster.");
```
Comment (Minecraftschurli) [inline src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMEnglishLanguageProvider.java]: `craterRadius` in text?

File: src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMEnglishLanguageProvider.java [hunk 2]
```diff
@@ -266,9 +266,9 @@ protected void addTranslations() {
...
         skillTranslation(AMSpellParts.EXPLOSION.getId(), "Explosion", "Creeper? Aww Man!", "components", "You can cause an explosion, destroying and dropping blocks around its center.");
-        skillTranslation(AMSpellParts.FALLING_STAR.getId(), "[WIP] Falling Star", "Shiny! Wait, is it falling towards me?", "components", "You can call down a star from the skies and cause it to strike all entities within the blast radius. It will harm friendly targets but not the caster.$(br2)There is a short delay between casting the spell and the impact. This spell will not work underground (the star will fall onto the surface).");
+        skillTranslation(AMSpellParts.FALLING_STAR.getId(), "[WIP] Falling Star", "Shiny! Wait, is it falling towards me?", "components", "You can call down a star from the skies and cause it to strike all entities within the blast craterRadius. It will harm friendly targets but not the caster.$(br2)There is a short delay between casting the spell and the impact. This spell will not work underground (the star will fall onto the surface).");
```
Comment (Minecraftschurli) [inline src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMEnglishLanguageProvider.java]: again

File: src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMEnglishLanguageProvider.java [hunk 3]
```diff
@@ -266,9 +266,9 @@ protected void addTranslations() {
...
         skillTranslation(AMSpellParts.FIRE_DAMAGE.getId(), "Fire Damage", "You shall burn!", "components", "With a word, you can release your will, and fire will erupt from out in front of you, searing everything in its path. Fire damage is hard hitting, but many nether mobs are immune to its effects.");
-        skillTranslation(AMSpellParts.FIRE_RAIN.getId(), "Fire Rain", "Well, at least I'm not going to catch a cold...", "components", "You have learned to summon a terrible firestorm, which will do large amounts of damage to all entities in its radius. Firestorm has a built-in $(l:shapes/aoe)AoE$() to it.$(br2)Firestorm does not ignite the ground and will never harm its caster.");
+        skillTranslation(AMSpellParts.FIRE_RAIN.getId(), "Fire Rain", "Well, at least I'm not going to catch a cold...", "components", "You have learned to summon a terrible firestorm, which will do large amounts of damage to all entities in its craterRadius. Firestorm has a built-in $(l:shapes/aoe)AoE$() to it.$(br2)Firestorm does not ignite the ground and will never harm its caster.");
```
Comment (Minecraftschurli) [inline src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMEnglishLanguageProvider.java]: again

File: src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMEnglishLanguageProvider.java [hunk 4]
```diff
@@ -337,7 +337,7 @@ protected void addTranslations() {
...
         skillTranslation(AMSpellParts.WAVE.getId(), "Wave", "You might not want to surf on this one...", "shapes", "You can project a wave of magic in front of you that rolls forward, applying its effect to everything in its path.");
-        skillTranslation(AMSpellParts.WIZARDS_AUTUMN.getId(), "Wizard's Autumn", "Leaves fall in autumn. But it's kind of a slow process...", "components", "You have learned to focus your digging magic into a small radius that directly affects leaves.$(br2)This component has a built-in $(l:shapes/aoe)AoE$() that can be modified with $(l:modifiers/range)Range$() modifiers.");
+        skillTranslation(AMSpellParts.WIZARDS_AUTUMN.getId(), "Wizard's Autumn", "Leaves fall in autumn. But it's kind of a slow process...", "components", "You have learned to focus your digging magic into a small craterRadius that directly affects leaves.$(br2)This component has a built-in $(l:shapes/aoe)AoE$() that can be modified with $(l:modifiers/range)Range$() modifiers.");
```
Comment (Minecraftschurli) [inline src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMEnglishLanguageProvider.java]: again

File: src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMPatchouliBookProvider.java [hunk 5]
```diff
@@ -72,7 +72,7 @@ protected void addBooks(Consumer<BookBuilder<?, ?, ?>> consumer) {
...
-                .addSimpleTextPage("- Block of Diamond: 11$(br)- Block of Emerald: 12$(br)- Block of Netherite: 13")
+                .addSimpleTextPage("- Block of Diamond: 11$(br)- Block of Emerald: 12$(br)- Block of Netherite: 13$(br)- $(l:blocks/ores#moonstone)Block of Moonstone$(): 14$(br)- $(l:blocks/ores#sunstone)Block of Sunstone$(): 15")
                 .addSimpleTextPage("- Wooden Planks (Overworld): 1$(br)- (Mossy) Cobblestone: 2$(br)- Cobbled Deepslate: 2$(br)- Andesite/Diorite/Granite: 2$(br)- (Red) Sandstone: 2$(br)- Bricks: 3$(br)- (Mossy) Stone Bricks: 3$(br)- Polished Deepslate: 3$(br)- Deepslate Bricks/Tiles: 3$(br)- Polished Andesite/Diorite/Granite: 3$(br)- Smooth (Red) Sandstone: 3$(br)- Cut Copper (and variants): 3", "Structure Materials")
                 .addSimpleTextPage("- All Prismarine Variants: 4$(br)- Wooden Planks (Nether): 4$(br)- Blackstone: 4$(br)- Quartz Block: 4$(br)- (Red) Nether Bricks: 5$(br)- Polished Blackstone (Bricks): 5$(br)- Smooth Quartz: 5$(br)- End Stone Bricks: 6$(br)- Purpur Block: 6")
```
Comment (Minecraftschurli) [inline src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMPatchouliBookProvider.java:77]: We should think about generating these lists

File: src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMSpellPartDataProvider.java [hunk 6]
```diff
@@ -62,7 +62,7 @@ protected void createSpellPartData() {
                 .withIngredient(new IngredientSpellIngredient(Ingredient.of(AMTags.Items.GEMS_TOPAZ), 1))
                 .withIngredient(new IngredientSpellIngredient(Ingredient.of(Items.CLOCK), 1))
                 .withIngredient(new IngredientSpellIngredient(Ingredient.of(AMItems.TARMA_ROOT.get()), 1))
-                //.withIngredient(new EtheriumSpellIngredient(Set.of(EtheriumType.DARK), 5000))
+                .withIngredient(new EtheriumSpellIngredient(Set.of(EtheriumType.DARK), 5000))
```
Comment (Minecraftschurli) [inline src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMSpellPartDataProvider.java]: dark etherium is not obtainable in survival yet

File: src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMSpellPartDataProvider.java [hunk 7]
```diff
@@ -83,7 +83,7 @@ protected void createSpellPartData() {
                 .withIngredient(new IngredientSpellIngredient(Ingredient.of(AMTags.Items.GEMS_CHIMERITE), 1))
                 .withIngredient(new IngredientSpellIngredient(Ingredient.of(Items.CLOCK), 1))
                 .withIngredient(new IngredientSpellIngredient(Ingredient.of(AMItems.TARMA_ROOT.get()), 1))
-                //.withIngredient(new EtheriumSpellIngredient(Set.of(EtheriumType.LIGHT), 5000))
+                .withIngredient(new EtheriumSpellIngredient(Set.of(EtheriumType.LIGHT), 5000))
```
Comment (Minecraftschurli) [inline src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMSpellPartDataProvider.java]: same for light

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/Config.java [hunk 8]
```diff
@@ -96,6 +97,10 @@ private Server(ForgeConfigSpec.Builder builder) {
                     .comment("The maximum number of shape groups allowed for new spells.")
                     .translation(TranslationConstants.CONFIG + "max_shape_groups")
                     .defineInRange("max_shape_groups", 5, 0, 5);
+            SUNSTONE_CHANCE_FROM_OBSIDIAN = builder
+                    .comment("The chance that a sunstone ore will appear when water and lava form obsidian. Set to 0 to disable.")
+                    .translation(TranslationConstants.CONFIG + "sunstone_chance_from_obsidian")
+                    .defineInRange("sunstone_chance_from_obsidian", 0.005, 0, 1);
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/Config.java]: why is this still here?

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/Config.java [hunk 9]
```diff
@@ -167,7 +172,7 @@ private Server(ForgeConfigSpec.Builder builder) {
                     .translation(TranslationConstants.CONFIG + "entities.dryad.bonemeal_chance")
                     .defineInRange("bonemeal_chance", 0.01, 0, 1);
             DRYAD_BONEMEAL_RADIUS = builder
-                    .comment("The radius of bonemeal application.")
+                    .comment("The craterRadius of bonemeal application.")
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/Config.java]: what is this?

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

File: src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMPatchouliBookProvider.java (hunk 5 above)
Comment (IchHabeHunger54) [reply src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMPatchouliBookProvider.java:77]: Would probably require a custom page type, as determining how many pages we need is painful. But yes, I agree.

Comment (Minecraftschurli) [reply src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMPatchouliBookProvider.java:77]: No I mean during datagen

Comment (IchHabeHunger54) [reply src/data/java/com/github/minecraftschurlimods/arsmagicalegacy/data/AMPatchouliBookProvider.java:77]: My point stands: Determining how many pages we need is difficult and potentially impossible.

**Review by Minecraftschurli — APPROVED** [review]

**Pull request merged**